### ISSUE TITLE
inline iterators don't accept statically-nil ref parameters

### ISSUE THREAD

**Issue by disruptek** [nim-lang/Nim#16076]
```nim
type
  R = ref object
    z: int

iterator foo(x: int; y: R = nil): int =
  if y == nil:
    yield x
  else:
    yield y.z

for b in foo(10): discard
```

yields a codegen error on 1.0, 1.2, 1.4, 1.5:

```
/home/adavidoff/.cache/nim/z_d/@mz.nim.c: In function ‘NimMainModule’:
/home/adavidoff/.cache/nim/z_d/@mz.nim.c:140:44: error: request for member ‘z’ in something not a structure or union
  140 |    b__NzQOw9bP039bPPaEKTSSfPjQ = (*NIM_NIL).z;
      |                                            ^
```

**Comment by disruptek** [comment]
I guess it's really just any static-nil argument that blows it up.

**Comment by disruptek** [comment]
I guess it's really just any static-nil argument that blows it up.

**Comment by ringabout** [comment]
```nim
type
  R = ref object
    z: int

iterator foo(x: int; y: R = nil): int {.inline.} =
  if y == nil:
    yield x
  else:
    yield y.z

proc main() =
  for b in foo(10):
    echo b

main()
```

Could you test it in latest devel?

After this PR https://github.com/nim-lang/Nim/pull/16032
I got:
```nim
D:\QQPCmgr\Desktop\Nim\bin\nim_temp.exe  c -r -d:release --nimcache:nimcache test.nim
D:\QQPCmgr\Desktop\Nim\test.nim(25, 7) template/generic instantiation from here
D:\QQPCmgr\Desktop\Nim\test.nim(22, 11) Error: nil dereference is not allowed
```

cc/ @cooldome

If I remove this:
```nim
    # if a != nil and a.kind == nkNilLit:
    #    localError(g.config, n.info, "nil dereference is not allowed")
```
It compiles and gives the right result: `10`

The C codes:

```c
b = (*((tyObject_RcolonObjectType___BIc6TY7S7R6PZ6xxakvyEg*) NIM_NIL)).z;
```

**Comment by disruptek** [comment]
Seems like the latest 1.4.1 works as expected.  The remaining versions fail with one error or another.

```nim
type
  R = ref object
    z: int

iterator foo(y: R = nil): int =
  if y == nil:
    yield 10
  else:
    yield y.z

for b in foo(): discard
```

**Comment by cooldome** [comment]
Yes, I have fixed it recently in 87d3e5331a7838aa255deca8a840c2080b5424ce

**Comment by ringabout** [comment]
No, it doesn't work in devel after disallowing nil dereference #16032
```nim
D:\QQPCmgr\Desktop\Nim\test.nim(25, 7) template/generic instantiation from here
D:\QQPCmgr\Desktop\Nim\test.nim(22, 11) Error: nil dereference is not allowed
```
